component/org.wso2.carbon.identity.dpop/src/main/java/org/wso2/carbon/identity/dpop/listener/OauthDPoPInterceptorHandlerProxy.java

```diff
@@ -0,0 +1,179 @@
+                }
+            }
+        } catch (InvalidOAuthClientException e) {
+            throw new IdentityOAuth2ClientException("invalid_client", "Invalid Client");
+        }
+    }
+
+    /**
+     * This method handles DPoP proof validation during pre token renewal.
+     *
+     * @param tokenReqDTO  OAuth2AccessTokenReqDTO.
+     * @param tokReqMsgCtx OAuthTokenReqMessageContext.
+     * @param params       Map of parameters.
+     * @throws IdentityOAuth2Exception If an error occurs while validating DPoP proof.
+     */
+    @Override
+    public void onPreTokenRenewal(OAuth2AccessTokenReqDTO tokenReqDTO, OAuthTokenReqMessageContext tokReqMsgCtx,
+                                  Map<String, Object> params) throws IdentityOAuth2Exception {
+
+        String consumerKey = tokenReqDTO.getClientId();
+        if (log.isDebugEnabled()) {
+            log.debug(String.format("DPoP proxy intercepted the token renewal request from the client : %s.",
+                    consumerKey));
+        }
+
+        // Check if the Refresh token is of DPoP type.
+        boolean isDPoPBinding = false;
+        TokenBinding tokenBinding =
+                tokenBindingTypeManagerDao.getTokenBinding(tokenReqDTO.getRefreshToken(), OAuth2Util.isHashEnabled());
+        if (tokenBinding != null &&
+                StringUtils.equalsIgnoreCase(DPoPConstants.DPOP_TOKEN_TYPE, tokenBinding.getBindingType())) {
+            isDPoPBinding = true;
+        }
+
+        String dPoPProof = DPoPHeaderValidator.getDPoPHeader(tokReqMsgCtx);
+        if (isDPoPBinding && StringUtils.isNotBlank(dPoPProof)) {
+            // If DPoP proof is provided, then it will be handled as a DPoP token request.
+            if (!DPoPHeaderValidator.isValidDPoP(dPoPProof, tokenReqDTO, tokReqMsgCtx)) {
+                if (log.isDebugEnabled()) {
+                    log.debug(String.format("DPoP proof validation failed for the application Id : %s.", consumerKey));
+                }
+                throw new IdentityOAuth2ClientException(DPoPConstants.INVALID_DPOP_PROOF,
+                        DPoPConstants.INVALID_DPOP_ERROR);
+            }
+            if (!tokReqMsgCtx.getTokenBinding().getBindingValue().equalsIgnoreCase(tokenBinding.getBindingValue())) {
+                log.error("DPoP proof thumbprint value of the public key is not equal to binding value from" +
+                        " the refresh token.");
+                throw new IdentityOAuth2ClientException(DPoPConstants.INVALID_DPOP_PROOF,
+                        DPoPConstants.INVALID_DPOP_ERROR);
+            }
+        } else if (isDPoPBinding) {
+            if (log.isDebugEnabled()) {
+                log.debug(String.format("Renewal request received without the DPoP proof from the application Id: %s.",
+                        consumerKey));
+            }
+
+            throw new IdentityOAuth2ClientException(DPoPConstants.INVALID_DPOP_PROOF, "DPoP proof is required.");
```

Comment:
Is the error_code should be INVALID_DPOP_PROOF or something similar to DPOP proof not found?